File: pset6/vigenere/vigenere.py
from sys import argv
from cs50 import get_string


def main():
    # Validate command
    validate_argv()
    # Prompt user for plain text
    plaintext = take_input()
    # Encrypt plain text using a given key
    encrypted = encrypt_vigenere(plaintext, argv[1])
    print("ciphertext:", encrypted)


def validate_argv():
    if len(argv) != 2 or not argv[1].isalpha():
        print("Usage: python vigenere.py k")
        exit(1)


def take_input():
    while True:
        text = get_string("plaintext: ")
        if text:
            return text


def encrypt_vigenere(text, key):
    encrypted = ""
    key_len = len(key)
    key_index = 0
    for c in text:
        if not c.isalpha():
            encrypted += c
        else:
            # Calculate key position
            key_position = ord(key[key_index]) - calculate_shift(key[key_index])
            # Calculate character shift
            shift = calculate_shift(c)
            # Calculate encrypted character position
            new_position = (ord(c) - shift + key_position) % 26 + shift
            encrypted += chr(new_position)
            # Calculate key_index for the next iteration
            key_index = (key_index + 1) % key_len
    return encrypted


def calculate_shift(c):
    return ord('a' if c.islower() else 'A')


if __name__ == "__main__":
    main()

File: pset2/vigenere/vigenere.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>

int is_valid_keyword(string);
string encipher_vigenere(string, string);

int main(int argc, string argv[])
{
    if (argc != 2 || !is_valid_keyword(argv[1]))
    {
        printf("Usage: ./vigenere k\n");
        return 1;
    }

    string keyword = argv[1];

    string text = get_string("plaintext: ");

    printf("ciphertext: %s\n", encipher_vigenere(text, keyword));

    return 0;
}

int is_valid_keyword(string keyword)
{
    for (int i = 0, n = strlen(keyword); i < n; i ++)
    {
        if (!isalpha(keyword[i]))
        {
            return 0;
        }
    }
    return 1;
}

string encipher_vigenere(string text, string keyword)
{
    char key;
    int shift, keyword_len = strlen(keyword);

    for (int i = 0, n = strlen(text), j = 0; i < n; i ++)
    {
        if (isalpha(text[i]))
        {
            // Set shifting value
            shift = isupper(text[i]) ? 'A' : 'a';
            // Find a key to use
            key = isupper(keyword[j]) ? keyword[j] - 'A' : keyword[j] - 'a';

            // Encipher, using key and shifting value
            text[i] = (text[i] - shift + key) % 26 + shift;

            // Set keyword character index to use for the next text character
            j ++;
            if (j == keyword_len)
            {
                j = 0;
            }
        }
    }
    return text;
}